cranelift/codegen/src/isa/riscv64/inst.isle

```diff
@@ -773,13 +773,46 @@
 (decl imm (Type u64) Reg)
 (extern constructor imm imm)
 
+;; Imm12 Rules
+
+(decl imm12_zero () Imm12)
+(rule
+  (imm12_zero)
+  (imm12_const 0))
+
+(decl imm12_const (i32) Imm12)
+(extern constructor imm12_const imm12_const)
```

Comment:
Random thought while I'm looking at this: constructors like these could be marked `pure`, I think. There's no need to change this if these are always used in impure contexts anyway though.